src/sparse_blas/sparse_blas_loader.cpp

```diff
@@ -0,0 +1,240 @@
+*
+*
+* SPDX-License-Identifier: Apache-2.0
+*******************************************************************************/
+
+#include "oneapi/mkl/sparse_blas/sparse_blas.hpp"
+
+#include "function_table_initializer.hpp"
+#include "sparse_blas/function_table.hpp"
+#include "oneapi/mkl/detail/get_device_id.hpp"
+
+namespace oneapi::mkl::sparse {
+
+static oneapi::mkl::detail::table_initializer<mkl::domain::sparse_blas,
+                                              sparse_blas_function_table_t>
+    function_tables;
+
+void init_matrix_handle(sycl::queue &queue, matrix_handle_t *handle) {
+    auto libkey = get_device_id(queue);
+    function_tables[libkey].init_matrix_handle(queue, handle);
+}
+
+sycl::event release_matrix_handle(sycl::queue &queue, matrix_handle_t *handle,
+                                  const std::vector<sycl::event> &dependencies) {
+    auto libkey = get_device_id(queue);
+    return function_tables[libkey].release_matrix_handle(queue, handle, dependencies);
+}
+
+#define DEFINE_SET_CSR_DATA(FP_TYPE, FP_SUFFIX, INT_TYPE, INT_SUFFIX)                          \
+    template <>                                                                                \
+    void set_csr_data(sycl::queue &queue, matrix_handle_t handle, INT_TYPE num_rows,           \
+                      INT_TYPE num_cols, index_base index, sycl::buffer<INT_TYPE, 1> &row_ptr, \
+                      sycl::buffer<INT_TYPE, 1> &col_ind, sycl::buffer<FP_TYPE, 1> &val) {     \
+        auto libkey = get_device_id(queue);                                                    \
+        function_tables[libkey].set_csr_data_buffer##FP_SUFFIX##INT_SUFFIX(                    \
+            queue, handle, num_rows, num_cols, index, row_ptr, col_ind, val);                  \
+    }                                                                                          \
+    template <>                                                                                \
+    sycl::event set_csr_data(sycl::queue &queue, matrix_handle_t handle, INT_TYPE num_rows,    \
+                             INT_TYPE num_cols, index_base index, INT_TYPE *row_ptr,           \
+                             INT_TYPE *col_ind, FP_TYPE *val,                                  \
+                             const std::vector<sycl::event> &dependencies) {                   \
+        auto libkey = get_device_id(queue);                                                    \
+        return function_tables[libkey].set_csr_data_usm##FP_SUFFIX##INT_SUFFIX(                \
+            queue, handle, num_rows, num_cols, index, row_ptr, col_ind, val, dependencies);    \
+    }
+
+DEFINE_SET_CSR_DATA(float, _sf, std::int32_t, _i32)
+DEFINE_SET_CSR_DATA(double, _sd, std::int32_t, _i32)
+DEFINE_SET_CSR_DATA(std::complex<float>, _cf, std::int32_t, _i32)
+DEFINE_SET_CSR_DATA(std::complex<double>, _cd, std::int32_t, _i32)
+DEFINE_SET_CSR_DATA(float, _sf, std::int64_t, _i64)
```

Comment:
Is this how all other domains are dealing with this or is it just in sparse BLAS that we have to deal with these combos and suffixes?

Reply:
I would suggest we add an additional macro layer here:

```cpp
#define FOR_EACH_INT_TYPE(_macro_forward)   \
        _macro_forward(std::int32_t, _i32); \
        _macro_forward(std::int64_t, _i64)

#define FOR_EACH_FP_TYPE(_macro_forward)          \
        _macro_forward(float, _rf);               \
        _macro_forward(double, _rd);              \
        _macro_forward(std::complex<float>, _cf); \
        _macro_forward(std::complex<double>, _cd)

#define FOR_EACH_FP_AND_INT_TYPE(_macro_forward)                       \
        _macro_forward(float, _rf, std::int32_t, _i32);                \
        _macro_forward(double, _rd, std::int32_t, _i32);               \
        _macro_forward(std::complex<float>, _cf, std::int32_t, _i32);  \
        _macro_forward(std::complex<double>, _cd, std::int32_t, _i32); \
        _macro_forward(float, _rf, std::int32_t, _i64);                \
        _macro_forward(double, _rd, std::int32_t, _i64);               \
        _macro_forward(std::complex<float>, _cf, std::int32_t, _i64);  \
        _macro_forward(std::complex<double>, _cd, std::int32_t, _i64)
```

The way to use it on APIs that contain both integer and fp types such as `set_csr_data` would be would be a single line:
```cpp
FOR_EACH_FP_AND_INT_TYPE(DEFINE_SET_CSR_DATA);
```
For APIs such as `GEMV` that only depend on fp types, you would use
```cpp
FOR_EACH_FP_TYPE(DEFINE_GEMV);
```
That should reduce a lot of your boilerplate code and reduce room for making mistakes in modifying existing APIs or adding new ones in the future.

Reply:
I have re-used this idea as of https://github.com/oneapi-src/oneMKL/pull/374/commits/3228b12622b8feb6caa109952d825e0cd65af171
I have applied the same kind of macro to `src/sparse_blas/sparse_blas_loader.cpp` and `src/sparse_blas/backends/mkl_common/mkl_helper.hpp`